### ISSUE TITLE
Automatic performance testing

### ISSUE THREAD

**Issue by milankl** [SpeedyWeather/SpeedyWeather.jl#464]
As also suggested by @maximilian-gelbrecht we should get some automated performance testing. Not necessarily a test that has to be passed for merging, but more like an automated way to record and easily see whether something makes the model considerably slower. Maybe similar to code coverage but for performance tests. If possible then even automatically written into a performance.md so that we can track performance across development too?

I'm imagining nothing too fancy just run the model with a few options like

- Barotropic, ShallowWater, PrimitiveDry, PrimitiveWet
- T31, T63, T127
- FullGaussianGrid, OctahedralGaussianGrid
- dynamics+physics, dynamics only, physics only
- Float32, Float64
- 8, 16 vertical levels, 1 for barotropic/shallowwater

@giordano do you suggest using [PkgBenchmark.jl](https://github.com/JuliaCI/PkgBenchmark.jl) for this?

**Comment by maximilian-gelbrecht** [comment]
Yeah, I was already thinking a little bit about it as well. I can also help to set up some of the test. 
In the end, it's also question how much CI time we have for that, isn't it? All possible combinations might be a bit much, but we should cover all important bits.

**Comment by milankl** [comment]
> But I guess, we could also define a small benchmark set that's executed at every commit and helps us to identify if any changes during development have an impact on the performance, and a larger benchmark set that is only manually triggered and presented in some way in the docs.

Yeah I think that's the tricky thing to figure out how to do this as automatic as possible. I'd be happy not to test the performance on every commit, but on the other hand I was thinking of not testing every combination of the options above but only relative to a (PrimitiveWet, T31L8, OctahedralGaussianGrid{Float32}, dynamics+physics)-default. So that would leave us with O(10) tests which is less than the continuous integration tests.

**Comment by giordano** [comment]
Yes, I use `PkgBenchmark.jl`, for example in [`Measurements.jl`](https://github.com/JuliaPhysics/Measurements.jl), you can see the setup at https://github.com/JuliaPhysics/Measurements.jl/tree/36aca91776832de0f458f719a51711923eff8be2/benchmark, and the CI job at https://github.com/JuliaPhysics/Measurements.jl/blob/36aca91776832de0f458f719a51711923eff8be2/.github/workflows/benchmark.yml.  The results of the benchmark runs are stored in the [`benchmark-results`](https://github.com/JuliaPhysics/Measurements.jl/tree/benchmark-results) branch.

Some caveats about this benchmark framework:

* if you use GitHub-hosted runners, be prepared to get varying results: each run (comparison between PR and target branch) is done on a single machine, but if you want to compare absolute numbers across different runs remember that may have been executed on different machines, and so with potentially different hardware (PkgBenchmark prints at the bottom of the page the hardware use, for reference)
* you need to check in the manifest that you want to use for running the benchmark, which means that if a PR you need to update some dependencies (e.g. need to change major version, or add new deps) then that benchmark run won't work, because you can't use a single manifest for both branches (PR branch, and target branch).  Following runs after the PR has been merged will be fine
* if you change code which affects the syntax of the benchmark functions, then that benchmark run won't work, because you can't have the same benchmarking  script for both branches (PR branch, and target branch).  Following runs after the PR has been merged will be fine.

These caveats may or may not be too annoying for you.  I don't have major troubles with `Measurements.jl` because it's a slow-moving package, but for other packages with higher activity, which may directly or indirectly affect the benchmarks as explained above, but it's good to be aware of this.  Honestly I don't think these caveats are completely unavoidable (maybe the first point could be avoided once https://github.com/JuliaLang/Pkg.jl/pull/3783 comes into effect).

**Comment by milankl** [comment]
Awesome, thanks @giordano that's fabulous input we can start working with.

**Comment by maximilian-gelbrecht** [comment]
I can have a go at it towards the end of the week I think

**Comment by milankl** [comment]
Thanks for checking that! I'd just circumvent the BenchmarkTools part in PkgBenchmarks if possible. If we ran a simulation like
```julia
julia> run!(simulation,period = Day(100))
Weather is speedy: 100%|██████████████████| Time: 0:00:02 ( 8.93 millenia/day)
```
Then one can just check the `model.feedback.progress_meter` with `tinit` being the unix time at the start of the simulation (*) and `tlast` the unix time afterwards. So the reported 9000years/day as speed ☝🏼 (which are a measured over the last 0.1s by default I believe, so not the full 2s the simulation ran for) can be recalculated from
```julia
julia> (;tlast, tinit) = model.feedback.progress_meter
julia> simulation.prognostic_variables.clock.period.value/(tlast-tinit)/365
8981.388564258898
```
`clock.period` is the `period` argument in the `run!` call (in `Seconds`) and so this gives you simulated seconds per wallclock seconds or equivalently days/day, and I just divided by 365 to get the typical SYPD unit. Most of our benchmarks will be easily measurable in SYPD so I'd stick to that.

(*) I just realised that the feedback is actually initialised before the `first_timesteps` which may include some compilation because it's the first time the `timestep!` is called, which calls all the RHS and it also recalculates the semi implicit terms. So I'd suggest that `initialize!(::Feedback,...)` should be called just before the main time loop `for i in 2:clock.n_timesteps` and the `finish!(::Feedback,...)` directly after to remove all the overhead from initialisation and finishing up things.

**Comment by maximilian-gelbrecht** [comment]
> I'd just circumvent the BenchmarkTools part in PkgBenchmarks if possible.

I don't think that's easily possible unfortunately. `PkgBenchmarks` expects a benchmark suite from `BenchmarkTools`. But maybe @giordano knows more? Maybe it's possible to overload/extend certain functions to configure it differently? It doesn't look very straightforward though, so maybe just doing it manually is easier.

**Comment by milankl** [comment]
> (*) I just realised that the feedback is actually initialised before the `first_timesteps` which may include some compilation because it's the first time the `timestep!` is called, which calls all the RHS and it also recalculates the semi implicit terms. So I'd suggest that `initialize!(::Feedback,...)` should be called just before the main time loop `for i in 2:clock.n_timesteps` and the `finish!(::Feedback,...)` directly after to remove all the overhead from initialisation and finishing up things.

This is part of #465
